onnx/version_converter/BaseConverter.h

```diff
@@ -0,0 +1,59 @@
+// Version converter interface for ONNX models between different opset versions.
+
+#pragma once
+
+#include "onnx/common/ir.h"
+#include "onnx/common/ir_pb_converter.h"
+#include "onnx/common/stl_backports.h"
+#include "onnx/proto_utils.h"
+#include "onnx/defs/schema.h"
+#include <utility>
+#include <iostream>
+// TODO: Remove this import once actual adapters are imported
+#include "onnx/version_converter/adapters/adapter.h"
+
+namespace ONNX_NAMESPACE { namespace version_conversion {
+
+struct BaseVersionConverter {
+  // Schema for adapters: {<op_name>:{<from_domain>$<from_version>:{<to_domain>
+  // <to_version>: adapter}}}
+  std::unordered_map<std::string, std::unordered_map<std::string, std::unordered_map<std::string, Adapter*>>> adapters;
+
+  std::unordered_map<Node*, OpSchema*> current_opschemas;
+
+  BaseVersionConverter() {}
+
+  ~BaseVersionConverter() {
+    for (auto op : adapters) {
+      for (auto init : op.second) {
+        for (auto target : init.second) {
+          target.second->~Adapter();
+        }
+      }
+    }
+  };
+
+  Adapter* adapter_lookup(Node* op,
+      const OpSetID& initial_version,
+      const OpSetID& target_version) {
+    // TODO: Abstract to helper?
+    barf("BaseConverter does not include an implementation of adapter_lookup.  "
+        "Please use a more specific converter, such as DefaultConverter.");
+  }
+
+  ONNX_NAMESPACE::ModelProto convert_version(
+      const ONNX_NAMESPACE::ModelProto& mp_in,
+      const OpSetID initial_version,
+      const OpSetID target_version) {
+    barf("BaseConverter does not include an implementation of convert_version.  "
+        "Please use a more specific converter, such as DefaultConverter.");
+  };
+
+  void registerAdapter(Adapter* a_ptr, std::string domain) {
+    OpSetID iv = a_ptr->initial_version;
```

Comment:
Taking these as const references doesn't need to copy it

Reply:
Even the pointer?

Reply:
I didn't mean the function arguments but the `OpSetId iv` and below. Function arguments might also be made const references if not retained, but as you noticed that does not make sense for the pointer.